### PR TITLE
stm32h7:SPI Add buffers for DMA & and FIX DMA

### PR DESCRIPTION
**Pull request by davids5**
This add an optional ability to have buffering in the SPI driver for DMA. 

If a SPI Bus uses an "in driver buffers" we will incur 2 copies,  The copy cost is << less the non DMA transfer time and having the buffer in the driver ensures DMA can be used. 

This is needed for because the SPI API does not support passing the buffer extent, so the only extent is buffer + the transfer size.  This is an issue on H7 with caching. These can sizes be less than the cache line size, and not aligned and typically greater then 4 bytes, which is about the break even point for the DMA IO overhead.

The SPI driver was violating the DMA/SPI setup as described in 49.4.14 Communication using DMA (direct memory addressing) of RM0433 Rev 6. The result was that the TX DMA operation would always return FIFO overrun/underrun (FEIFx) indicating a FIFO error event occurred on the stream.

This PR fixes this issue.

### CODE REVIEW

File: arch/arm/src/stm32h7/stm32_spi.c [hunk 1]
```diff
@@ -489,9 +577,14 @@ static struct stm32_spidev_s g_spi4dev =
 #ifdef CONFIG_STM32H7_SPI_INTERRUPTS
   .spiirq   = STM32_IRQ_SPI4,
 #endif
-#ifdef CONFIG_STM32H7_SPI_DMA
+#ifdef CONFIG_STM32H7_SPI3_DMA
```
Comment (davids5) [inline arch/arm/src/stm32h7/stm32_spi.c:580]:
```suggestion
#ifdef CONFIG_STM32H7_SPI4_DMA
```

**Comment by Ouss4** [comment]
@davids5  With a dummy RX transfer, aren't you invalidating more data from the cache then needed?
That's of course not harmful, but probably not what's desired.

File: arch/arm/src/stm32h7/stm32_spi.c [hunk 2]
```diff
@@ -1768,23 +1840,48 @@ static void spi_exchange(FAR struct spi_dev_s *dev, FAR const void *txbuffer,
...
+   * the buffer in the driver ensures DMA can be used. This is bacause
+   * the API does not support passing the buffer extent so the only
+   * extent is buffer + the transfer size. These can sizes be less than
+   * the cache line size, and not aligned and tyicaly greater then 4
+   * bytes, which is about the break even point for the DMA IO overhead.
+   */
+
+  if (txbuffer && priv->txbuf)
+    {
+      if (nbytes > priv->buflen)
```
Comment (xiaoxiang781216) [inline arch/arm/src/stm32h7/stm32_spi.c:1859]: Is it safe to truncatethe buffer?

Comment (davids5) [reply arch/arm/src/stm32h7/stm32_spi.c:1859]: The API does allow for returning an error.  We could change that but in real use the assumption is the allocation is the worst case for the system on that bus.

Comment (davids5) [reply arch/arm/src/stm32h7/stm32_spi.c:1859]: The error is further down. I will force push....

Comment (xiaoxiang781216) [reply arch/arm/src/stm32h7/stm32_spi.c:1859]: Instead of returning error, it's better to:
1.Send the buffer segment by segment or
2.Switch to PIO mode

**Comment by davids5** [comment]
> @davids5 With a dummy RX transfer, aren't you invalidating more data from the cache than needed?
> That's of course not harmful, but probably not what's desired.

I am not seeing it.  When rxbuffer is null - there is no invalidation.

File: arch/arm/src/stm32h7/stm32_spi.c (hunk 2 above)
Comment (davids5) [reply arch/arm/src/stm32h7/stm32_spi.c:1859]: @xiaoxiang781216 - thanks - force pushed using truncated value in DMA setup

Comment (davids5) [reply arch/arm/src/stm32h7/stm32_spi.c:1859]: @xiaoxiang781216 - I would be better. But this is for performance on small packets. The assumption is the worst case size is set for the bus. PIO would defeat the purpose. Looping would also.

**Pull request merged**